### PR TITLE
Add support for AssumeRole for DynamoDB locks and S3 remote state

### PR DESCRIPTION
**Pull request by npepinpe**
As it says, adds basic support for aws_role_arn (DynamoDB config) and role_arn (S3 remote state config), similar to the terraform AWS provider.

I realise it's not particularly well tested, but reading through the current tests, I was unsure about the way to go there. I'm nowhere near a Go expert, so I'll defer to your judgement, and gladly make any necessary adjustments.

### CODE REVIEW

File: aws_helper/config.go [hunk 1]
```diff
@@ -22,5 +25,29 @@ func CreateAwsSession(awsRegion, awsProfile string) (*session.Session, error) {
 		return nil, errors.WithStackTraceAndPrefix(err, "Error finding AWS credentials (did you set the AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY environment variables?)")
 	}
 
+	if awsRoleArn != "" {
+		err = assumeRole(session, awsRoleArn)
+		if err != nil {
```
Comment (brikis98) [inline aws_helper/config.go]: You can combine these two lines into one:

```go
if err := assumeRole(session, awsRoleArn); err != nil {
  ...
}
```

File: aws_helper/config.go [hunk 2]
```diff
@@ -22,5 +25,29 @@ func CreateAwsSession(awsRegion, awsProfile string) (*session.Session, error) {
 		return nil, errors.WithStackTraceAndPrefix(err, "Error finding AWS credentials (did you set the AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY environment variables?)")
 	}
 
+	if awsRoleArn != "" {
+		err = assumeRole(session, awsRoleArn)
```
Comment (brikis98) [inline aws_helper/config.go]: This method calls `Credentials.Get()` a few lines above, so you should probably handle the role stuff earlier. Moreover, I think it can be done simpler, without the need for an extra method:

```go
func CreateAwsSession(awsRegion, awsProfile, awsRoleArn string) (*session.Session, error) {
  session, err := // ...

  if awsRoleArn != "" {
    session.Config.Credentials = stscreds.NewCredentials(sess, assumeRoleArn)
  }

  _, err = session.Config.Credentials.Get()
  // ...
```

Comment (brikis98) [inline aws_helper/config.go]: BTW, `session` is a bad choice for a variable name here, as it shadows the `session` package that is imported. That has nothing to do with your code, just a screw up on my behalf. Could you rename it to something like `sess` instead to avoid confusion?

File: remote/remote_state_test.go [hunk 3]
```diff
@@ -13,10 +13,11 @@ func TestToTerraformRemoteConfigArgs(t *testing.T) {
...
 		Config: map[string]string{
-			"encrypt": "true",
-			"bucket":  "my-bucket",
-			"key":     "terraform.tfstate",
-			"region":  "us-east-1",
+			"encrypt":  "true",
+			"bucket":   "my-bucket",
+			"key":      "terraform.tfstate",
+			"region":   "us-east-1",
+			"role_arn": "arn:aws:iam::318125143396:role/delegated-admin.apps-sandbox",
```
Comment (brikis98) [inline remote/remote_state_test.go]: Is this a real ARN from your AWS account?

Comment (npepinpe) [inline remote/remote_state_test.go]: Oh wow, rookie mistake there, I was using this to test in a sandbox, my bad.

**Comment by npepinpe** [comment]
Some of the integration tests are failing at the moment, mostly related to includes, I'll update with the test output once I get them all to pass. I'm also not sure if to add integration tests for this case, as there are none currently for the profile either. Not sure what the right approach would be.

**Comment by brikis98** [comment]
> Some of the integration tests are failing at the moment, mostly related to includes, I'll update with the test output once I get them all to pass. 

Ah, let me know when you have that sorted.

> I'm also not sure if to add integration tests for this case, as there are none currently for the profile either. Not sure what the right approach would be.

It would be great to add integration tests for this, but you'd probably have to dynamically generate IAM roles and then clean them up after to make it work... I'm not sure on the best approach for that.

**Comment by brikis98** [comment]
@npepinpe This PR would need some serious rebasing, as Terragrunt no longer manages locking by itself since Terraform v0.9 came out. I'm going to close this one for now, but please reopen if you have a chance to revisit this stuff.
